---
title: "HW 13: Strings and things"
author: "Afreen Alam"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

***
#### Load libraries

```{r}
library(tidyverse)
library(here)
library(ggthemes)
library(gghighlight)
```

***
#### Part 1: DNA strings

**Import the data**

```{r}
beta_gliadin_raw <- read_lines(here("data", "beta_gliadin.txt"), skip = 1)
head(beta_gliadin_data)
```


**From many strings to one**


Replace all of the whitespace with nothing.

Replace all of the digits with nothing.

```{r}
(beta_data <- beta_gliadin_raw %>%
  str_replace_all(c(" " = "", "\\d" = "", "\\t" = "")))
```

Use str_c to collapse the strings. Print the variable to ensure you have just one string. Here are the first 100 characters.

```{r}
(beta_data_c <- beta_data %>%
   str_c(collapse = "")
   )
```

Convert it to upper case with str_to_upper().

```{r}
(beta_data_final <- beta_data_c %>%
   str_to_upper()
   )
```

 How long is the total DNA sequence? You should get 3310 characters.
 
```{r}
str_length(beta_data_final)
```

**Sequence analysis: find the promoter region**

Use str_count() and regex to count the number of IUPAC letters in the sequence that are not A, C, G, or T.
(Ans - None)

```{r}
str_count(beta_data_final, "[^ATCG]")
```

Count how many possible start codons are in the sequence.

```{r}
str_count(beta_data_final, "ATG")
```
Count how many possible stop codons are located in the sequence. For full credit, I’ll be more impressed if you search for both stop codons with a single regular expression. Hint: |

```{r}
str_count(beta_data_final, "(TGA|TAA)")
```

Assign the two motif sequences to variables caat_box and tata_box, respectively.

Count how many times each of these motifs occurs in the DNA sequence.

```{r}
caat_box <- "CCAAT"
tata_box <- "CTATAA"


# Number of caat_box
str_count(beta_data_final, caat_box)

# Number of tata_box
str_count(beta_data_final, tata_box)
```

Make a regex search string called caat_seq that begins with “CCAAT”, ends with “ATG”", and allows for a stretch of 100-150 nucleotides between the two. Hint: Remember {start, stop}.
Count how many times this possible sequence occurs.

```{r}
caat_seq <- "CCAAT.{100,150}ATG"
str_count(beta_data_final, caat_seq)
```

Make a regex search string called tata_seq that begins with “CTATAA”, ends with “ATG”, and allows for a stretch of 50-100 nucleotides between the two. Think about this carefully.

Count how many times this possible sequence occurs.

```{r}
tata_seq <- "CTATAA.{50,100}ATG"
str_count(beta_data_final, tata_seq)
```

Use str_extract_all to extract all of the caat_seq sequences and save them in a variable called caat_seqs_all. Use the simplify = TRUE argument in your function call (?str_extract_all for help).

```{r}
(caat_seqs_all <- beta_data_final %>%
   str_extract_all(caat_seq,
                   simplify = TRUE)
   )
```

How long is each extracted sequence? Make a mental note of the length of each one. You should have two sequences that differ in length by nine nucleotides.

```{r}
str_length(caat_seqs_all)
```
Use str_which() to find which caat_seq has the tata_box sequence. Use that value to extract the full promotor sequence into a variable called promotor.
Print promotor and determine it’s length

```{r}
# Position of which caat_seq has the tata_box sequence

str_which(caat_seqs_all, tata_box)

# Using position to extract sequence
(promotor <- caat_seqs_all[str_which(caat_seqs_all,
                                     tata_box)])

# Determine length
str_length(promotor)
```


***
#### Part 2: Ants

**Import the data**

```{r}
ants_data <- read_csv(here("data", "ants.csv"))
head(ants_data)
```


**Wrangle the data**

```{r}
(ants_data_wrangled <- ants_data %>%
   pivot_longer(names_to = "species",
                values_to = "abundance",
                matches("[A-Z].+\\.[a-z]+",
                        ignore.case = FALSE)) %>%
   select(-c(season:sample)) %>%
   mutate(species = str_replace(species,
                                "(.{4}).*\\.(.{4}).*", "\\1\\2"),
          year = factor(year,
                        ordered = TRUE,
                        levels = c(2010, 2011))) %>%
   group_by(year, species) %>%
   summarise(total_abundance = sum(abundance))
   )
```

**Plot the data**

Use the aes aesthetic in ggplot for mapping. Use x = reorder(species, total) to sort the species order from most to least abundant. total should be on the Y-axis.



```{r}
ggplot(ants_data_wrangled,
       aes(x = reorder(species,total_abundance),
           y = total_abundance),
       group = species) +
  geom_line(color = "gray30") +
  geom_point(aes(shape = year),
             size = 2) +
  coord_flip() +
  labs(x = NULL, y = "Total Abundance", shape = "Year") +
theme_minimal()  
```

***
#### Part 3: Featuring Phlox Flowers

**Import and wrangle the first data set**

```{r}
flower_size_data <- read_csv(here("data", "flower_size.csv"))
```



```{r}
(flower_size_wrangled <- flower_size_data %>%
   filter(!is.na(`Flower number`)) %>%
   select(species = Species,
          cor_length = `Corolla length (cm)`,
          throat_width = `Corolla width throat (cm)`) %>%
   separate(species, into = c("genus", "species"), extra = "drop") %>%
   group_by(genus, species) %>%
   summarise(mean_length = mean(cor_length, na.rm = TRUE),
             max_length = max(cor_length, na.rm = TRUE),
             mean_width = mean(throat_width, na.rm = TRUE),
             max_width = max(throat_width, na.rm = TRUE))
 )

```


**Import and wrangle the second data set.**

```{r}
pollinators_raw <- read_csv(here("data", "flower_pollinators.csv"))
```

```{r}
(pollinators <- pollinators_raw %>%
   select(-c(starts_with("Source"), Color)) %>%
   filter(!is.na(Pollinator)) %>%
   separate(Species, into = c("genus", "species"), extra = "drop") %>%
   separate(Pollinator, into = "Syndrome", extra = "drop", sep = "(,| )")
   )
```


**Join the datasets**

```{r}
(phlox <- flower_size_wrangled %>%
   inner_join(pollinators)
   )
```

**Plotting**

Make a scatterplot to show the relationship between mean_length (x-axis) and mean_width (y-axis). Color the points by Syndrome.

```{r}
phlox %>%
  ggplot(aes(mean_length, mean_width, color = Syndrome)) +
  geom_point() +
  labs(x = "Mean length", y = "Mean width")
```

Use mutate with str_replace() and regex patterns to replace all occurences of “beetle” with “beetles,” and “bee” with “bees”. Be careful. If you don’t do this right, then you could end up with “beetless”. Make use of the $ match character.

```{r}
phlox %>%
  mutate(Syndrome = str_replace(Syndrome, "(^be.*e$)", "\\1s")) %>%
  ggplot(aes(mean_length, mean_width, color = Syndrome)) +
  geom_point() +
  labs(x = "Mean length", y = "Mean width")
```

Once again, use mutate() with str_replace and a regex pattern to replace all instances of “beetles” or “flies” with “other insects”.

```{r}
phlox %>%
  mutate(Syndrome = str_replace(Syndrome, "(^be.*e$)", "\\1s"),
         Syndrome = str_replace(Syndrome, ".+l.?es", "other insects")) %>%
  ggplot(aes(mean_length, mean_width, color = Syndrome)) +
  geom_point() +
  labs(x = "Mean length", y = "Mean width")
```

Use mutate and case_when to create a new column called syndrome_group and to change autogamous to Autogamous (upper case A), bat to Mammal, hummingbird to Bird, and everything else to Insect. Creating a new column would allow you to access the individual syndromes for future plots, when necessary.

```{r}
(phlox_categorised <- phlox %>%
  mutate(Syndrome = str_replace(Syndrome, "(^be.*e$)", "\\1s"),
         Syndrome = str_replace(Syndrome, ".+l.?es", "other insects"),
         syndrome_group = case_when(
           Syndrome == "autogamous" ~ "Autogamous",
          Syndrome == "bat" ~ "Mammal",
          Syndrome == "hummingbird" ~ "Bird",
          TRUE ~ "Insect"
           
         )))
  
  
phlox_categorised %>%
  ggplot(aes(mean_length, mean_width, color = syndrome_group)) +
  geom_point() +
  labs(x = "Mean length", y = "Mean width", color = "Syndrome")
```


**Highlighting groups with gghighlight**

Change the above plot so that it highlights both the mammal and bird syndrome groups. 

```{r}
phlox_categorised %>%
  ggplot(aes(mean_length, mean_width, color = syndrome_group)) +
  geom_point() +
  labs(x = "Mean length", y = "Mean width", color = "Syndrome") +
  gghighlight(syndrome_group == "Mammal" | syndrome_group == "Bird",
              use_direct_label = FALSE)
```


























































